### PR TITLE
EDSC-4303: Refactor map tests

### PR DESCRIPTION
**Pull request by macrouch**
# Overview

### What is the feature?

Refactor the Playwright map tests to be more reliable.

### What is the Solution?

Split up single test into 4 different tests to make them easier to understand and work within.
Removed assertions of the individual lines drawn by Leaflet, we will rely on the Leaflet library to work as advertised. Instead we are only going to check the right number of shapes are drawn. This will remove the flaky-ness of 1-2 pixel changes between runs.

I also fixed a bug I found with uploading a shapefile in a polar projection. After the map switched to the new projection you could not remove the drawing of the shapefile from the map.

I also fixed a bug where selecting a granule on the map causes the granule outlines to move to a different part of the map.

### What areas of the application does this impact?

Map tests

# Testing

Run the playwright tests. Locally I am still seeing issues when running more than one browser at a time. The `act` tool is working well when limited to a single browser.

# Checklist

- [x] I have added automated tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

File: tests/e2e/map/map_shapefile.spec.js [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+
+        // Waiting for the URL to include the correct zoom level ensures the map is finished drawing
+        await page.waitForURL(/zoom=6/, { timeout: 3000 })
+
+        await expect(page.getByTestId('filter-stack-item__hint')).toHaveText('1 shape selected')
+
+        // The shapefile has 1 svg shapes that is auto selected, so 2 path elements
+        await expect(await page.locator('g path').all()).toHaveLength(2)
+
+        // Displays a modal
```
Comment (eudoroolivares2016) [inline tests/e2e/map/map_shapefile.spec.js:356]: We were closing the modal in the old one  // Close the modal
        await page.getByRole('button', { name: 'Close' }).click()
        I don't it was necessary

File: tests/e2e/paths/search/search.spec.js [hunk 2]
```diff
@@ -442,8 +446,9 @@ test.describe('Path /search', () => {
         await expect(page.getByTestId('spatial-display_shapefile-name')).toHaveText('test.geojson')
         await expect(page.getByTestId('filter-stack-item__hint')).toHaveText('1 shape selected')
 
-        // Test leaflet has drawn the shape correctly
-        await expect(page.locator('.leaflet-interactive').first()).toHaveAttribute('d', 'M1161 404L1122 480L1257 499L1161 404z')
+        // Test leaflet has drawn the shape
+        await expect(await page.locator('g path').first()).toBeVisible()
+        await expect(await page.locator('g path').all()).toHaveLength(2)
```
Comment (eudoroolivares2016) [inline tests/e2e/paths/search/search.spec.js:451]: This polygon has 2 g paths?

File: tests/e2e/map/map_spatial.spec.js [hunk 3]
```diff
@@ -0,0 +1,794 @@
...
+        await expect(page.getByTestId('spatial-display_southwest-point')).toHaveValue('-9.53964,42.1875')
+        await expect(page.getByTestId('spatial-display_northeast-point')).toHaveValue('4.5297,56.25')
+
+        // Checking that the right number of results are loaded ensures that the route
+        // was fulfilled correctly with the succesfull paramCheck
+        await expect(page.getByText('Showing 2 of 2 matching collections')).toBeVisible()
+      })
+    })
+
+    test.describe('When editing a bounding box', () => {
```
Comment (eudoroolivares2016) [inline tests/e2e/map/map_spatial.spec.js:532]: Great that you were even able to add these edits for the shape tests

**Review by eudoroolivares2016 — APPROVED** [review]

File: tests/e2e/paths/search/search.spec.js (hunk 2 above)
Comment (eudoroolivares2016) [reply tests/e2e/paths/search/search.spec.js:451]: Interesting it does render both `geojson-svg leaflet-interactive` and a regular `leaflet-interactive` resolving comment

File: tests/e2e/map/map_granules.spec.js [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+        await page.route(/autocomplete$/, async (route) => {
+          await route.fulfill({
+            json: { feed: { entry: [] } }
+          })
+        })
+
+        await page.goto(`search/granules?p=${conceptId}&pg[0][v]=f&pg[0][gsk]=-start_date&q=${conceptId}&polygon[0]=42.1875,-2.40647,42.1875,-9.43582,49.21875,-9.43582,42.1875,-2.40647&tl=1622520000!3!!`)
+      })
+
+      test('displays a hint about using a bounding box instead of polygon and an MBR on the map', async ({ page }) => {
```
[the remark this answers is not in the document]
Comment (macrouch) [reply tests/e2e/map/map_granules.spec.js:282]: This isn't testing granules drawn on the map, it is testing the MBR

File: tests/e2e/map/map_shapefile.spec.js [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+        // The shapefile has 4 svg shapes, so 4 path elements
+        await expect(await page.locator('g path').all()).toHaveLength(4)
+
+        // Select the circle
+        // Playwright has trouble clicking the shape because of the leaflet-control-container, dispatchEvent('click') works
+        await page.locator('g path').nth(3).dispatchEvent('click')
+
+        await expect(page.getByTestId('filter-stack-item__hint')).toHaveText('1 shape selected')
+
+        // After clicking a shape the map will have 5 path elements
```
[the remark this answers is not in the document]
Comment (macrouch) [reply tests/e2e/map/map_shapefile.spec.js:191]: Yep, I'll add that in

File: tests/e2e/map/map_shapefile.spec.js (hunk 1 above)
Comment (macrouch) [reply tests/e2e/map/map_shapefile.spec.js:356]: Right, I didn't see the point

File: tests/e2e/map/map_shapefile.spec.js [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+
+        // Upload the shapefile
+        await uploadShapefile(page, 'multiple_shapes.geojson')
+
+        // Populates the spatial display field
+        await expect(
+          page.getByTestId('filter-stack__spatial')
+            .locator('.filter-stack-item__secondary-title')
+        ).toHaveText('Shape File')
+
```
[the remark this answers is not in the document]
Comment (macrouch) [reply tests/e2e/map/map_shapefile.spec.js:175]: Right, I didn't see the point

File: tests/e2e/map/map_shapefile.spec.js [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+        await expect(page.getByTestId('spatial-display_shapefile-name')).toHaveText('test.geojson')
+        await expect(page.getByTestId('filter-stack-item__hint')).toHaveText('1 shape selected')
+
+        // Checking that the right number of results are loaded ensures that the route
+        // was fulfilled correctly with the succesfull paramCheck
+        await expect(page.getByText('Showing 2 of 2 matching collections')).toBeVisible()
+
+        // Waiting for the URL to include the correct zoom level ensures the map is finished drawing
+        await page.waitForTimeout(100)
```
[the remark this answers is not in the document]
Comment (macrouch) [reply tests/e2e/map/map_shapefile.spec.js:71]: At one point I had a bunch of these in, I meant to remove them

File: tests/e2e/paths/search/search.spec.js (hunk 2 above)
Comment (macrouch) [reply tests/e2e/paths/search/search.spec.js:451]: The 2 paths come from the shapefile polygon being drawn, and the highlighted shape from the shape being selected

File: tests/e2e/map/map_granules.spec.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (eudoroolivares2016) [reply tests/e2e/map/map_granules.spec.js:282]: Ah I see now I thought MBR was referring to something else resolving

**Review by daniel-zamora — APPROVED** [review]

**Comment by daniel-zamora** [comment]
Looked through it yesterday, just didn't check the approval in case others wanted to take a look.  Approving now.

**Pull request merged**